### PR TITLE
[shopsys] move ready category seo mix from project-base to packages

### PR DESCRIPTION
**Pull request by grossmannmartin**
#### Description, the reason for the PR

ready category seo mix related code was moved from project-base to packages

#### Fixes issues
... 

#### Check the appropriate checkboxes below, please

- [x] [BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/) 
- [ ] New feature 
- [x] I have read and signed [License Agreement for contributions](https://www.shopsys.com/license-agreement)

----
:globe_with_meridians: Live Preview:
  - https://mg-move-seo-mix.odin.shopsys.cloud
  - https://cz.mg-move-seo-mix.odin.shopsys.cloud

### CODE REVIEW

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlMatcher.php [hunk 1]
```diff
@@ -32,22 +37,89 @@ public function match($pathinfo, RouteCollection $routeCollection, DomainConfig
             throw new ResourceNotFoundException();
         }
 
+        $matchedParameters = [];
+
+        if ($friendlyUrl->getRedirectTo() !== null) {
+            $matchedParameters['_route'] = $friendlyUrl->getRouteName();
+            $matchedParameters['_controller'] = 'Symfony\Bundle\FrameworkBundle\Controller\RedirectController::urlRedirectAction';
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlMatcher.php]: Maybe we could concat the action name with `RedirectController::class` to be a little bit more protected against a possible Symfony naming change?

Comment (vitek-rostislav) [inline packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlMatcher.php]: Maybe we could concat the action name with `RedirectController::class` to be a little bit more protected against a possible Symfony naming change?

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlMatcher.php [hunk 2]
```diff
@@ -32,22 +37,89 @@ public function match($pathinfo, RouteCollection $routeCollection, DomainConfig
...
+        if ($friendlyUrl->getRouteName() === 'front_category_seo') {
+            return $this->getMatchedParametersForMainFrontCategorySeoFriendlyUrl($friendlyUrl, $matchedParameters);
+        }
+
         $matchedParameters['_route'] = $friendlyUrl->getRouteName();
         $matchedParameters['id'] = $friendlyUrl->getEntityId();
 
         if (!$friendlyUrl->isMain()) {
-            $matchedParameters['_controller'] = 'Symfony\Bundle\FrameworkBundle\Controller\RedirectController::redirectAction';
+            $matchedParameters['_controller'] = 'Symfony\Bundle\FrameworkBundle\Controller\RedirectController::urlRedirectAction';
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlMatcher.php]: same here

File: packages/framework/src/Controller/Admin/CategoryController.php [hunk 3]
```diff
@@ -141,15 +147,17 @@ public function listAction(Request $request): Response
 
         return $this->render('@ShopsysFramework/Admin/Content/Category/list.html.twig', [
             'categoriesWithPreloadedChildren' => $categoriesWithPreloadedChildren,
-            'domainFilterNamespace' => $domainFilterNamespace,
             'isForAllDomains' => ($selectedDomainId === null),
+            'domainFilterNamespace' => $domainFilterNamespace,
+            'disabledFormFields' => $this->formBuilderHelper->hasFormDisabledFields(),
+            'getAllCategoryIdsInSeoMixes' => $this->categorySeoMixFacade->getAllCategoryIdsInSeoMixes(),
```
Comment (vitek-rostislav) [inline packages/framework/src/Controller/Admin/CategoryController.php]: nicpick, `getAllCategoryIdsInSeoMixes` is not a common name for a variable, sounds like a method name

File: packages/framework/src/Model/CategorySeo/DeleteReadyCategorySeoMixFacade.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    public function __construct(
+        protected readonly EntityManagerInterface $em,
+        protected readonly ReadyCategorySeoMixRepository $readyCategorySeoMixRepository,
+    ) {
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Parameter\Parameter $parameter
+     */
+    public function deleteAllWithParameter(Parameter $parameter): void
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/CategorySeo/DeleteReadyCategorySeoMixFacade.php:25]: one note to myself - this should be covered by upgrade notes (the method was extracted from `ReadyCategorySeoMixFacade`

File: packages/framework/src/Resources/views/Admin/Content/Category/list.html.twig [hunk 5]
```diff
@@ -1,49 +1,36 @@
...
-{% block title %}- {{ 'Categories'|trans }}{% endblock %}
-{% block h1 %}{{ 'Categories'|trans }}{% endblock %}
-{% block btn %}
-    <div class="wrap-bar__buttons">
-        <a href="{{ url('admin_category_new') }}" class="btn btn--primary btn--plus wrap-bar__btn">
-            <i class="btn__icon">+</i>
-            {{ 'Create new category'|trans }}
-        </a>
-    </div>
-{% endblock %}
```
Comment (vitek-rostislav) [inline packages/framework/src/Resources/views/Admin/Content/Category/list.html.twig]: this should not be removed, right? Category list headline is gone as well as button for creating a new category :open_mouth: 
image

vs 16.0 version:
image

File: packages/framework/src/Resources/views/Admin/Content/CategorySeo/newCombinations.html.twig [hunk 6]
```diff
@@ -40,7 +40,7 @@
             {% endif %}
 
             <td>
-                {{ render(controller('App\\Controller\\Admin\\CategorySeoController::readyCombinationButtonAction', {
+                {{ render(controller('Shopsys\\FrameworkBundle\\Controller\\Admin\\CategorySeoController::readyCombinationButtonAction', {
```
Comment (vitek-rostislav) [inline packages/framework/src/Resources/views/Admin/Content/CategorySeo/newCombinations.html.twig:43]: in this template (and in the others moved) is in vain used exclamation mark in the extend part in the first line (`{% extends '@!...` should be actually `{% extends '@...`)

File: packages/frontend-api/src/Model/Product/Filter/ProductFilterFacade.php [hunk 7]
```diff
@@ -183,4 +184,39 @@ public function getProductFilterConfigForSearch(string $searchText): ProductFilt
...
+     * @param \Shopsys\FrameworkBundle\Model\Product\Flag\Flag $flag
+     * @return \Shopsys\FrameworkBundle\Model\Product\Filter\ProductFilterConfig
+     */
+    public function getProductFilterConfigForFlag(Flag $flag): ProductFilterConfig
+    {
+        $locale = $this->domain->getLocale();
+        $cacheKey = sprintf('flag_%s_%s', $locale, $flag->getId());
+
+        if (!array_key_exists($cacheKey, $this->productFilterConfigCache)) {
+            $this->productFilterConfigCache[$cacheKey] = $this->productFilterConfigFactory->createForFlag(
```
Comment (vitek-rostislav) [inline packages/frontend-api/src/Model/Product/Filter/ProductFilterFacade.php]: optional, in a separate commit, we should use our new `InMemoryCache` here, right?

File: packages/frontend-api/src/Model/Resolver/Category/CategoryQuery.php [hunk 8]
```diff
@@ -77,4 +104,163 @@ protected function getVisibleOnDomainAndSlug(string $urlSlug): Category
...
+        return $this->categoryFacade->getCategoriesOfProductByFilterData($productFilterData);
+    }
+
+    /**
+     * @param \GraphQL\Type\Definition\ResolveInfo $info
+     * @param string|null $uuid
+     * @param string|null $urlSlug
+     * @return \Shopsys\FrameworkBundle\Model\Category\Category|\Shopsys\FrameworkBundle\Model\CategorySeo\ReadyCategorySeoMix
+     */
+    public function categoryOrSeoMixByUuidOrUrlSlugQuery(
```
Comment (vitek-rostislav) [inline packages/frontend-api/src/Model/Resolver/Category/CategoryQuery.php:89]: We still have `project-base/app/config/graphql/types/Query/Category/CategoryQuery.types.yaml` - its content should be moved to the fe API package while `categoryByUuidOrUrlSlugQuery` should be removed from this class

File: project-base/app/src/FrontendApi/Resolver/Image/Exception/ImageTypeInvalidUserError.php [hunk 9]
```diff
@@ -1,21 +0,0 @@
-<?php
-
-declare(strict_types=1);
-
-namespace App\FrontendApi\Resolver\Image\Exception;
-
-use Shopsys\FrontendApiBundle\Model\Error\EntityNotFoundUserError;
-use Shopsys\FrontendApiBundle\Model\Error\UserErrorWithCodeInterface;
-
-class ImageTypeInvalidUserError extends EntityNotFoundUserError implements UserErrorWithCodeInterface
```
Comment (vitek-rostislav) [inline project-base/app/src/FrontendApi/Resolver/Image/Exception/ImageTypeInvalidUserError.php:10]: it seems like `ImageTypeInvalidUserError` is not used anywhere, could you please check it and remove it as well if possible?

Comment (vitek-rostislav) [inline project-base/app/src/FrontendApi/Resolver/Image/Exception/ImageTypeInvalidUserError.php:10]: it is out of the scope, however, it seems like `Shopsys\FrontendApiBundle\Model\Resolver\Image\Exception\ImageTypeInvalidUserError` is not used anywhere, could you please check it and remove it as well if possible? (also the "image-type-invalid" string occurrences)

File: project-base/app/tests/FrontendApiBundle/Functional/Category/ReadyCategorySeoMix/ReadyCategorySeoMixTest.php [hunk 10]
```diff
@@ -1,424 +0,0 @@
...
-use App\Model\Product\Flag\Flag;
-use App\Model\Product\Parameter\Parameter;
-use App\Model\Product\Parameter\ParameterFacade;
-use Shopsys\FrameworkBundle\Component\ArrayUtils\ArraySorter;
-use Shopsys\FrameworkBundle\Component\Translation\Translator;
-use Shopsys\FrameworkBundle\Model\CategorySeo\ReadyCategorySeoMix;
-use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
-use Tests\FrontendApiBundle\Test\GraphQlTestCase;
-
-class ReadyCategorySeoMixTest extends GraphQlTestCase
```
Comment (vitek-rostislav) [inline project-base/app/tests/FrontendApiBundle/Functional/Category/ReadyCategorySeoMix/ReadyCategorySeoMixTest.php:21]: maybe we can preserve these tests "just" by replacing the "slug" queries with "categoryByUuidOrUrlSlugQuery"  category queries

File: packages/framework/src/Model/Product/Flag/FlagRepository.php [hunk 11]
```diff
@@ -137,4 +137,65 @@ public function getFlagIdsByUuids(array $flagUuids): array
...
+        if ($flag === null) {
+            throw new FlagNotFoundException(sprintf('Flag with ID "%s" does not exist.', $flagId));
+        }
+
+        return $flag;
+    }
+
+    /**
+     * @param string $locale
+     * @return \App\Model\Product\Flag\Flag[]
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Product/Flag/FlagRepository.php]: uncaught App namespace here

File: packages/framework/src/Model/Product/Flag/FlagRepository.php [hunk 12]
```diff
@@ -137,4 +137,65 @@ public function getFlagIdsByUuids(array $flagUuids): array
...
+            ->orderBy(OrderByCollationHelper::createOrderByForLocale('ft.name', $locale), 'asc')
+            ->setParameter('locale', $locale);
+
+        return $flagsQueryBuilder->getQuery()->getResult();
+    }
+
+    /**
+     * @param string $uuid
+     * @param string $locale
+     * @return \App\Model\Product\Flag\Flag
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Product/Flag/FlagRepository.php]: uncaught App namespace here

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php [hunk 13]
```diff
@@ -162,6 +162,7 @@ public function productsWithOverlyingEntityQuery(Argument $argument, ResolveInfo
         }
 
         if ($argument['flagSlug'] !== null) {
+            /** @var \App\Model\Product\Flag\Flag $flag */
```
Comment (vitek-rostislav) [inline project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php]: uncaught App namespace here

Comment (vitek-rostislav) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php]: (in the later commit, this code is moved to the fe API package but the inline annotation is preserved

File: packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php [hunk 14]
```diff
@@ -226,4 +205,106 @@ function ($offset, $limit) use ($argument, $category, $productFilterData, $order
...
+            $category = $this->categoryQuery->categoryOrSeoMixByUuidOrUrlSlugQuery($info, null, $argument['categorySlug']);
+
+            return $this->productsByCategoryOrReadyCategorySeoMixQuery(
+                $argument,
+                $category,
+            );
+        }
+
+        if ($argument['brandSlug'] !== null) {
+            /** @var \App\Model\Product\Brand\Brand $brand */
```
Comment (vitek-rostislav) [inline packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php]: app namespace

File: packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php [hunk 15]
```diff
@@ -226,4 +205,106 @@ function ($offset, $limit) use ($argument, $category, $productFilterData, $order
...
+            $brand = $this->brandQuery->brandByUuidOrUrlSlugQuery(null, $argument['brandSlug']);
+
+            return $this->productsByBrandQuery(
+                $argument,
+                $brand,
+            );
+        }
+
+        if ($argument['flagSlug'] !== null) {
+            /** @var \App\Model\Product\Flag\Flag $flag */
```
Comment (vitek-rostislav) [inline packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php]: app namespace

File: packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php [hunk 16]
```diff
@@ -226,4 +205,106 @@ function ($offset, $limit) use ($argument, $category, $productFilterData, $order
...
+            $batchLoadDataId,
+        );
+    }
+
+    /**
+     * @param \Overblog\GraphQLBundle\Definition\Argument $argument
+     * @param \GraphQL\Type\Definition\ResolveInfo $info
+     * @return \Shopsys\FrontendApiBundle\Model\Product\Connection\ProductConnection|\GraphQL\Executor\Promise\Promise
+     */
+    public function productsWithOverlyingEntityQuery(Argument $argument, ResolveInfo $info): ProductConnection|Promise
```
Comment (vitek-rostislav) [inline packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php:256]: there is still `project-base/app/config/graphql/types/Query/Product/ProductQuery.types.yaml` where this query is defined (and overrides the `products` query from `QueryDecorator`, could you move that as well, please?

File: upgrade-notes/backend_20241022_082619.md [hunk 17]
```diff
@@ -0,0 +1,27 @@
+#### use moved Ready Category Seo Mix code from packages ([#3494](https://github.com/shopsys/shopsys/pull/3494))
+
+-   [features moved](#movement-of-features-from-project-base-to-packages) from project-base to the frontend-api and framework packages:
```
Comment (vitek-rostislav) [inline upgrade-notes/backend_20241022_082619.md:3]: as I mentioned somewhere at the beginning of the review, we should IMHO also mention the new `DeleteReadyCategorySeoMixFacade` with extracted `deleteAllWithParameter()` method

**Review by vitek-rostislav — CHANGES_REQUESTED** [review]
Hats off, I love the outcome! :heart_eyes: 
image
awesome!

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlMatcher.php (hunk 1 above)
Comment (grossmannmartin) [reply packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlMatcher.php]: good idea! changed in 76d970376f

Comment (grossmannmartin) [reply packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlMatcher.php]: good idea! changed in 76d970376f

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlMatcher.php (hunk 2 above)
Comment (grossmannmartin) [reply packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlMatcher.php]: good idea! changed in 76d970376f

File: packages/framework/src/Controller/Admin/CategoryController.php (hunk 3 above)
Comment (grossmannmartin) [reply packages/framework/src/Controller/Admin/CategoryController.php]: good 575d31edb6

File: packages/framework/src/Model/CategorySeo/DeleteReadyCategorySeoMixFacade.php (hunk 4 above)
Comment (grossmannmartin) [reply packages/framework/src/Model/CategorySeo/DeleteReadyCategorySeoMixFacade.php:25]: added in 0ea2e69645

File: upgrade-notes/backend_20241022_082619.md (hunk 17 above)
Comment (grossmannmartin) [reply upgrade-notes/backend_20241022_082619.md:3]: added in 0ea2e69645

File: packages/framework/src/Resources/views/Admin/Content/Category/list.html.twig (hunk 5 above)
Comment (grossmannmartin) [reply packages/framework/src/Resources/views/Admin/Content/Category/list.html.twig]: 😅  of course, it should not be removed... thanks for pointing it  out - fixed in b5a8c73c77

File: packages/framework/src/Resources/views/Admin/Content/CategorySeo/newCombinations.html.twig (hunk 6 above)
Comment (grossmannmartin) [reply packages/framework/src/Resources/views/Admin/Content/CategorySeo/newCombinations.html.twig:43]: Oh, cool. I did not notice - f1d415889f

File: packages/frontend-api/src/Model/Product/Filter/ProductFilterFacade.php (hunk 7 above)
Comment (grossmannmartin) [reply packages/frontend-api/src/Model/Product/Filter/ProductFilterFacade.php]: cool idea 7518bfb6f3

File: project-base/app/src/FrontendApi/Resolver/Image/Exception/ImageTypeInvalidUserError.php (hunk 9 above)
Comment (grossmannmartin) [reply project-base/app/src/FrontendApi/Resolver/Image/Exception/ImageTypeInvalidUserError.php:10]: removed a503582752

Comment (grossmannmartin) [reply project-base/app/src/FrontendApi/Resolver/Image/Exception/ImageTypeInvalidUserError.php:10]: removed a503582752

File: packages/framework/src/Model/Product/Flag/FlagRepository.php (hunk 11 above)
Comment (grossmannmartin) [reply packages/framework/src/Model/Product/Flag/FlagRepository.php]: cool, thanks - 37e9e89ecf

File: packages/framework/src/Model/Product/Flag/FlagRepository.php (hunk 12 above)
Comment (grossmannmartin) [reply packages/framework/src/Model/Product/Flag/FlagRepository.php]: cool, thanks - 37e9e89ecf

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php (hunk 13 above)
Comment (grossmannmartin) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php]: a9e191a9a0

File: packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php (hunk 14 above)
Comment (grossmannmartin) [reply packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php]: a9e191a9a0

File: packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php (hunk 15 above)
Comment (grossmannmartin) [reply packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php]: a9e191a9a0

File: packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php (hunk 16 above)
Comment (grossmannmartin) [reply packages/frontend-api/src/Model/Resolver/Products/ProductsQuery.php:256]: moved in c2a8cb8461

File: project-base/app/tests/FrontendApiBundle/Functional/Category/ReadyCategorySeoMix/ReadyCategorySeoMixTest.php (hunk 10 above)
Comment (grossmannmartin) [reply project-base/app/tests/FrontendApiBundle/Functional/Category/ReadyCategorySeoMix/ReadyCategorySeoMixTest.php:21]: wow, great idea – tests are back in 415e8637fc

File: packages/frontend-api/src/Model/Resolver/Category/CategoryQuery.php (hunk 8 above)
Comment (grossmannmartin) [reply packages/frontend-api/src/Model/Resolver/Category/CategoryQuery.php:89]: done in 938e035b39 ;) nice

**Review by vitek-rostislav — APPROVED** [review]

**Pull request merged**